Add support for alternate boost-context backends

elfprince13 opened the pull request:
The fcontext backend is the default high-performance backend.
The ucontext backend is needed when using Boost context in conjunction with ASAN.
The WinFibers backend...also exists.

cf. https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/cc/implementations__fcontext_t__ucontext_t_and_winfiber.html

elfprince13 committed:
Add support for alternate boost-context backends

The fcontext backend is the default high-performance backend.
The ucontext backend is needed when using Boost context in conjunction with ASAN.
The WinFibers backend...also exists.

cf. https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/cc/implementations__fcontext_t__ucontext_t_and_winfiber.html

tldahlgren reviewed (approved):
LGTM.  Deferring merge decision to maintainers/package experts.

tldahlgren commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -156,6 +156,13 @@ def libs(self):
 
+    variant('context-impl',
+            default='fcontext',
+            values=('fcontext','ucontext','winfib'),
```
Please correct style issues:

```suggestion
            values=('fcontext', 'ucontext', 'winfib'),
```

elfprince13 replied:
Sorted in [65d7b34](https://github.com/spack/spack/pull/30496/commits/65d7b34ce1839e576f377bba64a5eb9c256a2a58)

tldahlgren reviewed (changes requested):
Please correct the style error; otherwise, LGTM.

hainest commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -156,6 +156,13 @@ def libs(self):
+            values=('fcontext','ucontext','winfib'),
+            multi=False,
+            description='Use the specified backend for boost-context',
+            when='+context')
```
What version of Boost introduced this feature? I'm working to make sure variants are correctly mapped to versions where they are actually available.

elfprince13 replied:
The relevant pages are here:
https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/stack/sanitizers.html and here https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/cc/implementations__fcontext_t__ucontext_t_and_winfiber.html

It's available at least as far back as 1.65, haven't gone further back.

hainest replied:
I'll run my automated tests and check.  **edit**: We don't need to hold this PR for this.

elfprince13 replied:
Incidentally there's an option for supporting valgrind that would be great to get rolled in as well (as a separate PR):

https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/stack/valgrind.html

hainest replied:
Oh, very nice! I will try to see about getting at least a reference list of the features/properties.

hainest commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -481,6 +488,11 @@ def determine_b2_options(self, spec, options):
+        context_impl = spec.variants['context-impl'].value
+        if '+context' in spec and context_impl in {'ucontext', 'winfib'}:
+            options.extend(['context-impl=%s' % context_impl])
+
```
Assuming b2 won't raise any fuss if we pass `context-impl=fcontext`, there's no need to check for it here. That will remove this maintenance hazard in keeping the check here and the values of the variant in synch.

```suggestion
if '+context' in spec:
  context_impl = spec.variants['context-impl'].value
  options.append('context-impl=%s' % context_impl)
```

elfprince13 replied:
The documentation page I linked above only discusses the possibility of `ucontext` or `winfib` as arguments.

hainest replied:
In 1.79.0,

``` console
$ ./bootstrap.sh --with-libraries=context

$ ./b2 context-impl=moo
/boost_1_79_0/tools/build/src/build/feature.jam:491: in validate-value-string from module feature
error: "moo" is not a known value of feature <context-impl>
error: legal values: "fcontext" "ucontext" "winfib"

$ ./b2 context-impl=fcontext
  ...
    - context                  : building
  ...
The Boost C++ Libraries were successfully built!
```

Let's assume for now that always works. I'll add it to my TODO list to double-check.

elfprince13 replied:
Cool - i just verified `fcontext` is present as an option at least as far back as 1.65:
https://github.com/boostorg/context/blob/boost-1.65.0/build/Jamfile.v2

I'll go ahead and remove that extra check.

I will note that the special-casing for the defines introduced in [65d7b34](https://github.com/spack/spack/pull/30496/commits/65d7b34ce1839e576f377bba64a5eb9c256a2a58) should be kept as `BOOST_USE_FCONTEXT` is not used anywhere.

hainest replied:
Oh, that's... neat. Particularly since ucontext was [removed](https://www.boost.org/doc/libs/1_79_0/libs/context/doc/html/context/rationale/other_apis_.html#ucontext) in POSIX.1-2008. Boost, why did you do that?!

Out of curiosity, which backend are you using?

elfprince13 replied:
fcontext is definitely the sensible default for most use-cases, but ucontext is the only way to cooperate with ASAN if you're doing memory debugging (or was - it seems to have broken at some point recently since ASAN promptly reported a bug in boost itself when I switched)

hainest replied:
Oh sorry, yes, you mentioned that in your commit. It's interesting ASAN is relying on the deprecated/removed interface. I would assume it's really difficult to instrument the fcontext usage.

hainest reviewed (changes requested):
It would be great if we could get a list of the properties that b2 supports, but I couldn't find a listing in the [docs](https://www.boost.org/doc/libs/1_79_0/tools/build/doc/html/index.html#bbv2.reference.features). How did you find this one?

elfprince13 committed:
Fixups for code review on #30496

Whitespace fixes.
Also improve the client interface somewhat by setting the appropriate flags in the build environment.

tldahlgren reviewed (dismissed):
Removing my blocking review related to style check.

elfprince13 committed:
Simplify context-impl variant logic

Also adds explanatory comments for the special-casing that remains.

hainest commented:
Looks like the pipeline is stuck.

@spackbot re-run pipeline
